aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,312 @@
+
+.. image:: filters/_static/add-header.png
+
+The available built-in attribute-based filters are:
+	- `ActionFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ActionFilterAttribute/index.html>`_
+	- `AuthorizationFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/AuthorizationFilterAttribute/index.html>`_
+	- `ExceptionFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ExceptionFilterAttribute/index.html>`_
+	- `ResultFilterAttribute <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/Filters/ResultFilterAttribute/index.html>`_
+
+Dependency Injection
+^^^^^^^^^^^^^^^^^^^^
+Filters are not directly returned from :doc:`dependency injection </fundamentals/dependency-injection>` (DI). Filters that are also attributes, and which are added directly to controller classes or action methods cannot have constructor dependencies provided by DI - their constructor properties must be specified when they are declared. This is a limitation of how attributes work. However, if your filters have dependencies you need to access from DI, there are several supported approaches. You can apply your filter using either the `ServiceFilter <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/ServiceFilterAttribute/index.html>`_ or `TypeFilter <https://docs.asp.net/projects/api/en/latest/autoapi/Microsoft/AspNet/Mvc/TypeFilterAttribute/index.html>`_ attribute, or you can implement IFilterFactory on your attribute, or you can access the services you require from the context parameter provided to your filter's On\ *Stage* methods.
+
+For an example of this last technique, consider the following method, which needs to access a service that has been registered with DI. The service is accessible from the context parameter:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/OrderLoggingActionFilterAttribute.cs
+  :language: c#
+  :emphasize-lines: 3-4
+  :lines: 10-20
+  :dedent: 8
+
+Cancellation and Short Circuiting
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+You can short-circuit the filter pipeline at any point by setting the ``Result`` property on the context parameter provided to the filter method. For instance, the following ``ShortCircuitingResourceFilter`` will prevent any other filters from running later in the pipeline, including any action filters.
+
+.. _short-circuiting-resource-filter:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/ShortCircuitingResourceFilterAttribute.cs
+  :language: c#
+  :emphasize-lines: 12-15
+
+When this filter is applied to an action that also would have an action filter applied (later in the pipeline), such as the ``SomeResource`` action below, the action filter is not executed (in this case the ``AddHeader`` action filter is applied at the class level, but the behavior would be the same if it were applied to the method).
```

Comment:
In the following code, both the short circuit and ``AddHeader`` action filter act on the ``someresource`` action. Because the short circuit filter runs first and always short circuits, the ``AddHeader`` action filter never runs on the ``someresource`` action.  The behavior would be the same if both filters were applied to the action method. [ditto for both at the controller level - right?]